features/events/traps/src/main/java/org/opennms/netmgt/trapd/TrapReceiverSnmp4jImpl.java

```diff
@@ -150,19 +182,203 @@ public void stop() throws InterruptedException {
+            LOG.debug("Actual receive buffer size is {}", m_dgSock.getReceiveBufferSize());
+        } catch (SocketException e) {
+            LOG.info("Failed to set the receive buffer to {}", Integer.MAX_VALUE, e);
+        }
+
+        try {
+            LOG.debug("Opening datagram socket");
+            if (m_config.getListenAddress() != null && m_config.getListenAddress().length() != 0) {
+                m_dgSock.bind(new InetSocketAddress(InetAddressUtils.addr(m_config.getListenAddress()), m_config.getSyslogPort()));
+            } else {
+                m_dgSock.bind(new InetSocketAddress(m_config.getSyslogPort()));
             }
-        } catch (Throwable e) {
-            LOG.error("Handler execution failed in {}", this.getClass().getSimpleName(), e);
+        } catch (SocketException e) {
+            LOG.info("Failed to open datagram socket", e);
         }
+
+        // set to avoid numerous tracing message
+        boolean ioInterrupted = false;
+
+        // Construct one mutable {@link DatagramPacket} that will be used for receiving syslog messages 
+        DatagramPacket pkt = new DatagramPacket(buffer, length);
+
+        // now start processing incoming requests
+        while (!m_stop) {
+            if (m_context.isInterrupted()) {
+                LOG.debug("Thread context interrupted");
+                break;
+            }
+
+            try {
+                if (!ioInterrupted) {
+                    LOG.debug("Waiting on a datagram to arrive");
+                }
+
+                m_dgSock.receive(pkt);
+
+                SyslogConnection connection = new SyslogConnection(pkt, m_config);
+
+                try {
+                    for (TrapNotificationHandler handler : m_trapNotificationHandlers) {
+                      //  handler.handleTrapNotification(connection);
+                    }
+                } catch (Throwable e) {
+                    LOG.error("Handler execution failed in {}", this.getClass().getSimpleName(), e);
+                }
+
+                ioInterrupted = false; // reset the flag
+            } catch (SocketTimeoutException e) {
+                ioInterrupted = true;
+                continue;
+            } catch (InterruptedIOException e) {
+                ioInterrupted = true;
+                continue;
+            } catch (IOException e) {
+                if (m_stop) {
+                    // A SocketException can be thrown during normal shutdown so log as debug
+                    LOG.debug("Shutting down the datagram receipt port: " + e.getMessage());
+                } else {
+                    LOG.error("An I/O exception occured on the datagram receipt port, exiting", e);
+                }
+                break;
+            }
+
+        } // end while status OK
+
+        LOG.debug("Thread context exiting");
+
     }
 
+	@Override
+	public void trapReceived(TrapNotification trapNotification) {
+		m_backlogQ.submit(m_processorFactory.getInstance(trapNotification));
```

Comment:
Instead of doing this, you want to submit the TrapNotification to all of the handlers:
```
try {
  for (TrapNotificationHandler handler : m_trapNotificationHandlers) {
    handler.handleTrapNotification(connection);
  }
} catch (Throwable e) {
  LOG.error("Handler execution failed in {}", this.getClass().getSimpleName(), e);
}
```